samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -187,92 +174,83 @@ class KafkaCheckpointManager(
-    retryBackoff.run[Unit](
-      loop => {
-        val consumer = getConsumer()
-
-        val topicAndPartition = new TopicAndPartition(checkpointTopic, 0)
-
-        try {
-          var offset = startingOffset.getOrElse(getEarliestOffset(consumer, topicAndPartition))
-
-          info("Got offset %s for topic %s and partition 0. Attempting to fetch messages for %s." format(offset, checkpointTopic, entryType))
-
-          val latestOffset = getOffset(consumer, topicAndPartition, OffsetRequest.LatestTime)
 
-          info("Get latest offset %s for topic %s and partition 0." format(latestOffset, checkpointTopic))
-
-          if (offset < 0) {
-            info("Got offset 0 (no messages in %s) for topic %s and partition 0, so returning empty collection. If you expected the checkpoint topic to have messages, you're probably going to lose data." format (entryType, checkpointTopic))
-            return
-          }
-
-          while (offset < latestOffset) {
-            val request = new FetchRequestBuilder()
-              .addFetch(checkpointTopic, 0, offset, fetchSize)
-              .maxWait(500)
-              .minBytes(1)
-              .clientId(clientId)
-              .build
-
-            val fetchResponse = consumer.fetch(request)
-            if (fetchResponse.hasError) {
-              warn("Got error code from broker for %s: %s" format(checkpointTopic, fetchResponse.errorCode(checkpointTopic, 0)))
-              val errorCode = fetchResponse.errorCode(checkpointTopic, 0)
-              if (ErrorMapping.OffsetOutOfRangeCode.equals(errorCode)) {
-                warn("Got an offset out of range exception while getting last entry in %s for topic %s and partition 0, so returning a null offset to the KafkaConsumer. Let it decide what to do based on its autooffset.reset setting." format (entryType, checkpointTopic))
-                return
-              }
-              KafkaUtil.maybeThrowException(errorCode)
+    val ssp: SystemStreamPartition = new SystemStreamPartition(systemName, checkpointTopic, new Partition(0))
+    val systemConsumer = getSystemConsumer()
+    val (oldestOffset, newestOffset) = getEarliestOffset(checkpointTopic, new Partition(0))
+    systemConsumer.register(ssp, oldestOffset); // checkpoint stream should always be read from the beginning
+    systemConsumer.start();
+
+    var count = 0
+    try {
+      var done = false
+      var currentOffset = -1L
+      val newestOffsetLong = newestOffset.toLong
+      while (currentOffset < newestOffsetLong) {
+        // Map[SystemStreamPartition, List[IncomingMessageEnvelope]]
+        val envelopes = systemConsumer.poll(Collections.singleton(ssp), 1000)
+
+        val messages: util.List[IncomingMessageEnvelope] = envelopes.get(ssp)
+        val messagesSize = if (messages != null) messages.size else -1
+        info("CheckpointMgr read %s envelopes (%s messages) from ssp %s"
+                     format (envelopes.size(), messagesSize, ssp))
```

Comment:
Why is envelopes.size useful to log? Wouldn't it always be one given that we are polling for a single SSP?

Reply:
@vjagadish1989 From SystemConsumer:
"If no messages are available for a SystemStreamPartition that was supplied in the polling set, the map will not contain a key for the SystemStreamPartition."

However, messages == null sounds like a bug.

Also, s/messagesSize/numMessages

Reply:
wait. 'envelopes' can be empty, but if it is empty, then 'messages' will be null. If 'messages' is not null, it still theoretically could be empty (although that would be weird).